src/tqec/compile/compile.py

```diff
@@ -0,0 +1,229 @@
+
+    def scale_to(self, k: int) -> None:
+        """Scale the compiled graph to the given scale `k`."""
+        for tiled_blocks in self.tiles_by_time:
+            tiled_blocks.scale_to(k)
+        self._check_equal_block_size()
+
+    @property
+    def block_size(self) -> int:
+        self._check_equal_block_size()
+        return self.tiles_by_time[0].block_size
+
+    def generate_stim_circuit(
+        self,
+        k: int,
+        noise_models: Iterable[BaseNoiseModel] = (),
+    ) -> stim.Circuit:
+        """Generate the stim circuit from the compiled graph.
+
+        Args:
+            k: The scale factor of the templates.
+            noise_models: The noise models to be applied to the circuit.
+
+            A compiled stim circuit.
+        """
+        cirq_circuit = self.generate_cirq_circuit(k, noise_models)
+        cirq_circuit = transform_to_stimcirq_compatible(cirq_circuit)
+        return stimcirq.cirq_circuit_to_stim_circuit(cirq_circuit)
+
+    def generate_cirq_circuit(
+        self,
+        k: int,
+        noise_models: Iterable[BaseNoiseModel] = (),
+    ) -> cirq.Circuit:
+        """Generate the cirq circuit from the compiled graph.
+
+        Args:
+            k: The scale factor of the templates.
+            noise_models: The noise models to be applied to the circuit.
+
+        Returns:
+            A compiled cirq circuit.
+        """
+        self.scale_to(k)
+        # assemble circuits time slice by time slice, layer by layer.
+        circuits: list[list[cirq.Circuit]] = []
+        for tiles in self.tiles_by_time:
+            circuits.append(
+                [tiles.instantiate_layer(i) for i in range(tiles.num_layers)]
+            )
+        # construct observables
+        inplace_add_observables(circuits, self.observables, self.block_size)
+        # shift and merge circuits
+        circuit = self._shift_and_merge_circuits(circuits)
+        # apply noise models
+        for noise_model in noise_models:
+            circuit = circuit.with_noise(noise_model)
+        return circuit
+
+    def _shift_and_merge_circuits(
+        self, circuits: list[list[cirq.Circuit]]
+    ) -> cirq.Circuit:
+        circuits_by_time = []
+        for t, tiles in enumerate(self.tiles_by_time):
+            circuits_at_t = []
+            # We need to shift the circuit based on the shift of the tiled template.
+            origin_shift = tiles.tiled_template.origin_shift
+            for i in range(tiles.num_layers):
+                circuit_before_shift = circuits[t][i]
+                # Need to use `qubit_map` on `ScheduledCircuit` instead of
+                # `transform_qubits` on `cirq.Circuit` to map the MeasurementKey correctly.
+                scheduled = ScheduledCircuit(circuit_before_shift)
+                qubit_map = {
+                    q: q
+                    + (
+                        origin_shift.y * self.block_size * 2,
+                        origin_shift.x * self.block_size * 2,
```

Comment:
I do not understand the `* 2` here.

Reply:
Yes I think it's a mistake. It did not present itself as a bug in logical cnot circuit because all the `TiledBlocks` has origin at `(0,0)` in that model.